### PR TITLE
Use the hidden attribute & `display:none` to hide x-collapse content from the accessibility tree

### PR DESCRIPTION
**Pull request by philwolstenholme**
Big fan of Alpine! Thanks for all the time that has been put into it :) 

I was checking out [x-collapse](https://alpinejs.dev/plugins/collapse) and noticed that visually collapsed content is not hidden/removed from the accessibility tree, so keyboard users can tab to focusable elements that should be hidden, and the content may be announced to screenreader users.

At the moment we just hide content with `overflow: hidden; height: 0px;` but this isn't enough to hide content from assistive tech.

One way to hide the content safely would be to add a `hidden` attribute after the transition has ended and the content has been collapsed (or add it on load), and remove the attribute before un-collapsing the content.

The `hidden` attribute approach is particularly nice as it will play nicely with Tailwind's https://tailwindcss.com/docs/space utility. The utility won't add space around content that has been hidden with `hidden`:

```css
.space-y-1>:not([hidden])~:not([hidden]) {
    --tw-space-y-reverse: 0;
    margin-top: calc(.25rem * calc(1 - var(--tw-space-y-reverse)));
    margin-bottom: calc(.25rem * var(--tw-space-y-reverse))
}
```

`hidden` is essentially the same as `style="display:none"` except by using the `hidden` attribute we get that extra compatibility with Tailwind's space utilities 👍 

The browser user-agent stylesheet manages the hiding for us via a rule like:

```css
[hidden] { display: none; }
``` 

A downside of that is that author CSS can very easily override the `[hidden]` rule by setting a `display` property on the element being collapsed/uncollapsed, but that can be worked around by the author adding something like:

```css
[hidden] { display: none !important; }
```

This PR uses the `hidden` attribute and updates the Cypress tests for `x-collapse` to test for the presence or lack of the attribute.

### CODE REVIEW

**Comment by philwolstenholme** [comment]
I'm now wondering if it might be worth adding the `hidden` attribute _and_ an inline `display:none;` rule. The `hidden` attribute would help out Tailwind `space` utility class users, and the inline `display:none;` rule would be useful for preventing accessibility bugs where someone has set `display: flex;` (for example) on the element being toggled… It seems like duplication of effort but I also think including both approaches makes sense… 🤔

**Comment by SimoTod** [comment]
despite `hidden` looking very cool at first, it's unfortunately really weak and any styling rule would override it, as you said. x-show uses `display: none` so it may be a good idea to keep it consistent. 
Probably you can ask directly on the PR so Caleb can give you a feedback when he reviews.

File: packages/collapse/src/index.js
Comment (philwolstenholme) [inline packages/collapse/src/index.js:8]: I think I'll add a `display: none` rule here too, for the reasons given in the PR comments and here: https://github.com/alpinejs/alpine/pull/2353#issuecomment-965709781

File: packages/collapse/src/index.js
Comment (philwolstenholme) [inline packages/collapse/src/index.js:14]: For these changes, it'd be more consistent with `x-show` if we only set `el.style.display = 'none'`.

I've added the `el.hidden` parts to try and be helpful to Tailwind users who might be using the [space between](https://tailwindcss.com/docs/space) utilities to add spacing between collapsed elements. Without using `el.hidden`, the CSS would add extra space around collapsed items, which would be annoying. With `el.hidden`, Tailwind will ignore collapsed items. I don't know for sure, but this might be a CSS pattern used in other codebases too?

Comment (SimoTod) [reply packages/collapse/src/index.js:14]: Instead of reading display none, you can probably remove line 7 (which removes the style set by x-show)

Comment (philwolstenholme) [reply packages/collapse/src/index.js]: Ah yes, that's smarter! I did wonder about what the `el.style.removeProperty()` call was for, but I forgot that `x-collapse` works on top of `x-show`. I've just pushed a change with your suggestion.

**Comment by calebporzio** [comment]
Your points regarding the focusing nested elements are valid. The tailwind spacer utilities as well.

Because you're setting `style.display = null` here, will that conflict with a utility class like "flex"? or will it just remove it and default back to display flex?

If that's the case, I feel good about it.

Thank you as always for the feedback and help @SimoTod ❤️

**Comment by philwolstenholme** [comment]
Thanks for the quick response @calebporzio !

> Because you're setting `style.display = null` here, will that conflict with a utility class like "flex"? or will it just remove it and default back to display flex?

`style.display = null` will remove the `display: none` rule from the inline `style` attribute on the element, but any other CSS rules (like one coming from a `.flex` utility class) will be unaffected, and the element will go back to being styled by the CSS cascade rather than the inline style 👍 

You mentioning utilities has reminded me that if someone was using `x-collapse` with a utility that set the `display` property (e.g. `.flex`, `.inline` etc) and that utility was `!important` (e.g. via the Tailwind [important option](https://tailwindcss.com/docs/configuration#important)) then the utility would override the inline `display: none` coming from this PR, and the site would have the same accessibility bug of being able to focus collapsed content. Many users would never notice because visually it would all look fine thanks to the `height: 0; overflow: hidden` rule.

We could force our inline `display: none` to be `!important` too by using `el.style.setProperty('display', 'none', 'important');` perhaps, but that change would need to be made in `x-show` as well. I can't think of a time when I'd want to use `x-show` or `x-collapse` and _not_ have the extra certainty that my content will _definitely_ be hidden when it's meant to be, but you never know… What do you think @calebporzio and @SimoTod ?

File: tests/cypress/integration/plugins/collapse.spec.js
Comment (chrisrowe) [inline tests/cypress/integration/plugins/collapse.spec.js:7]:
```suggestion
            <h1 x-show="expanded" x-collapse>contents <a href="#">focusable content</a></h1>
```

**Review by chrisrowe — CHANGES_REQUESTED** [review]
Fix double closing link in test

**Comment by SimoTod** [comment]
It won't apply to collapse and i don't know if anyone uses it that way but you can force a div to be always visible on, for instance, mobile using important and have it showable on desktop. 
But as soon as you add transitions, it will start looking weird so I doubt anyone does it. 

As far as I know, nobody complained about x-show needing important so we can maybe wait.

About hidden, I'm not sure the spacing thing makes much difference because most of the time your element will be a block. Inline blocks, when disappearing, will make the rest of the line move which is probably not desirable.

**Comment by philwolstenholme** [comment]
I think maybe let's leave things as they are currently in the PR. There's a risk that `!important` utilities could undo the inline `display:none;` provided this fix, but at least then the accessibility issue is coming from the user's code (where they can fix it themselves) rather than coming from Alpine where it'll be harder for many users to fix themselves.

**Comment by SimoTod** [comment]
Yeah, If people start reporting it, we can add important at a later stage. Obviously the final decision is with Caleb so let's wait for his feedback

**Comment by calebporzio** [comment]
Ok, thanks for talking that through for me and for weighing in @SimoTod!

Merging this ❤️

**Pull request merged**
